Run partial CI jobs only for JS/SCSS/PHP file changes

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

Parts of #1047 

- Only run **JavaScript and CSS Linting** and **JavaScript Unit Tests** jobs when the change files are including `.js` or `.scss` files in pull requests, or in the commits on the primary branches.
- Only run **PHP Coding Standards** and **PHP Unit Tests** jobs when the change files are including `.php` files in pull requests, or in the commits on the primary branches.

Ref: https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#onpushpull_requestpull_request_targetpathspaths-ignore

### Screenshots:

#### 📷  Without any js/scss/php changes

![2022-03-14 17 49 45](https://user-images.githubusercontent.com/17420811/158148656-e5bca07d-a549-4797-aa2c-ffbdf595f04a.png)

#### 📷  Adding js file changes

![2022-03-14 17 52 58](https://user-images.githubusercontent.com/17420811/158148672-9724058b-0eae-49bb-b3e5-a0e35758bc8d.png)

#### 📷  Adding php file changes

![2022-03-14 17 54 20](https://user-images.githubusercontent.com/17420811/158148725-4425ad95-770e-4b7a-a881-eb1fe2a1c8e5.png)

### Detailed test instructions:

Check the three commits triggering jobs below, they should trigger more jobs as more files are changed.

💡  If change the order to commit PHP file changes first, it should expect only the PHP-related jobs to be added to trigger. But I think it should be ok to provide one of the verification flow in this PR.

### Changelog entry

eason9487 committed:
Only run JavaScript/CSS Linting and Unit Tests jobs when the change files are including js/scss files.

eason9487 committed:
Only run PHP Coding Standards and Unit Tests jobs when the change files are including php files.

eason9487 committed:
Remove an outdated code comment in a js file.

eason9487 committed:
Change two links in the code comments of PHP files to use https protocol.

puntope reviewed (commented):
Thanks for this PR. It would be much more efficient now with the paths. 

❓  I'm wondering what would happen if we are loading some kind of .svg file in the frontend and for example, we change the name of that .svg. Then this test will not run and we could potentially miss the error.

❓ A bit of the same happens with .json files in PHP. What if someone changes some mock.json file or `.js` file in the PHP test folder? 

Wondering if it'd better to add as well everything in ´/js/src´ for JS jobs and everything in `/src` and `/tests` for PHP jobs

eason9487 commented:
@puntope, thanks for the review comments.

> ❓ I'm wondering what would happen if we are loading some kind of .svg file in the frontend and for example, we change the name of that .svg. Then this test will not run and we could potentially miss the error.

If the name of an SVG file is changed, usually the js files importing that SVG file would be changed accordingly. So in this case, I assumed the js related jobs would be triggered in most cases.

If the only change is to the content of an SVG file, then the js related jobs will not be triggered. But it's not a concern since the `.svg` files are mocked when running unit tests. They should not affect the results of unit tests.

https://github.com/woocommerce/google-listings-and-ads/blob/1fbbde745aef5670290e08da462da000eb7836da/jest.config.js#L7-L8

> ❓ A bit of the same happens with .json files in PHP. What if someone changes some mock.json file or `.js` file in the PHP test folder?

Sorry, I'm not 100% sure of your concern. I assumed you meant the js/json files under the `/tests/mocks` and `/tests/proxy` folders. If yes, these files are not used in PHP unit tests.

Apart from the two mocking js files under `/tests/data`, I don't find any json or js files are associated with PHP unit tests. And it should be ok to not include those mocking js files.

---

But maybe I should add composer.json, package.json, and their lock files to the `paths` of job configs accordingly. 🤔

eason9487 committed:
Add package files and their lock files to unit test jobs

eason9487 commented:
Additional changes:
- Added composer.json and composer.lock to the `paths` of **PHP Unit Tests** job.
- Added package.json and package-lock.json to the `paths` of **JavaScript Unit Tests** job.

puntope reviewed (approved):
✅  LGTM

After discussion of some points that could affect this PR I think we are good to go.
